idea/idea-frontend-independent/src/org/jetbrains/kotlin/idea/quickfix/AddModifierFix.kt

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.quickfix
+
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import com.intellij.psi.util.PsiTreeUtil
+import org.jetbrains.kotlin.idea.KotlinBundleIndependent
+import org.jetbrains.kotlin.idea.inspections.KotlinUniversalQuickFix
+import org.jetbrains.kotlin.lexer.KtModifierKeywordToken
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.containingClass
+import org.jetbrains.kotlin.psi.psiUtil.containingClassOrObject
+
+open class AddModifierFix(
+    element: KtModifierListOwner,
+    protected val modifier: KtModifierKeywordToken
+) : KotlinCrossLanguageQuickFixAction<KtModifierListOwner>(element), KotlinUniversalQuickFix {
+    override fun getText(): String {
+        val element = element ?: return ""
+        if (modifier in modalityModifiers || modifier in KtTokens.VISIBILITY_MODIFIERS || modifier == KtTokens.CONST_KEYWORD) {
+            return KotlinBundleIndependent.message("fix.add.modifier.text", RemoveModifierFix.getElementName(element), modifier.value)
+        }
+        return KotlinBundleIndependent.message("fix.add.modifier.text.generic", modifier.value)
+    }
+
+    override fun getFamilyName() = KotlinBundleIndependent.message("fix.add.modifier.family")
+
+    protected fun invokeOnElement(element: KtModifierListOwner?) {
+        element?.addModifier(modifier)
+
+        if (modifier == KtTokens.ABSTRACT_KEYWORD && (element is KtProperty || element is KtNamedFunction)) {
+            element.containingClass()?.run {
+                if (!hasModifier(KtTokens.ABSTRACT_KEYWORD) && !hasModifier(KtTokens.SEALED_KEYWORD)) {
+                    addModifier(KtTokens.ABSTRACT_KEYWORD)
+                }
+            }
+        }
+    }
+
+    override fun invokeImpl(project: Project, editor: Editor?, file: PsiFile) {
+        val originalElement = element
+        invokeOnElement(originalElement)
+    }
+
+    override fun isAvailableImpl(project: Project, editor: Editor?, file: PsiFile): Boolean = element != null
```

Comment:
Makes sense that `invokeImpl` was changed from the original implementation since that has the MPP-checking parts, but why was this one changed too?

Reply:
Perhaps same point (or slightly different), but I think this one should be migrated from `AddModifierFixMpp` to here and `AddModifierFixMpp` won't need its own override?

Actually, it depends on when/how `isAvailableImpl` (and `invokeImpl` above) are used.

Reply:
The `canRefactor` depends on some FE1.0 logic that seems impossible to port over easily :(

Reply:
Okay, then maybe leave a TODO here to not forget or specify the intention? I assume `[AddModifierFixMpp#isAvailableImpl]` is not MPP specific, and thus should be here, once `canRefactor` is migrated/reimplemented?

Reply:
+1 to the suggestion to sprinkle this file with TODOs. 🙂